lib/codegen/src/verifier/mod.rs

```diff
@@ -82,23 +82,77 @@ pub use self::cssa::verify_cssa;
 pub use self::liveness::verify_liveness;
 pub use self::locations::verify_locations;
 
-// Create an `Err` variant of `VerifierResult<X>` from a location and `format!` arguments.
-macro_rules! err {
-    ( $loc:expr, $msg:expr ) => {
-        Err(::verifier::VerifierError {
+/// Report an error.
+macro_rules! report {
+    ( $errors: expr, $loc: expr, $msg: tt ) => {
+        $errors.0.push(::verifier::VerifierError {
             location: $loc.into(),
             message: String::from($msg),
         })
     };
 
-    ( $loc:expr, $fmt:expr, $( $arg:expr ),+ ) => {
-        Err(::verifier::VerifierError {
+    ( $errors: expr, $loc: expr, $fmt: tt, $( $arg: expr ),+ ) => {
+        $errors.0.push(::verifier::VerifierError {
             location: $loc.into(),
             message: format!( $fmt, $( $arg ),+ ),
         })
     };
 }
 
+/// Diagnose a fatal error, and return `Err` right away.
+macro_rules! fatal {
+    ( $( $arg: expr ),+ ) => ({
+        report!( $( $arg ),+ );
+        Err(())
+    });
+}
+
+/// Diagnose a non-fatal error, and return `Ok` right away.
+macro_rules! nonfatal {
+    ( $( $arg: expr ),+ ) => ({
+        report!( $( $arg ),+ );
+        Ok(())
+    });
+}
+
+/// Return `Ok(())` if no errors were reported in the given vector, and
+/// `Err(())` otherwise.
+macro_rules! ok_if_no_error {
+    ($errors:expr) => {
+        if $errors.0.is_empty() {
+            Ok(())
+        } else {
+            Err(())
+        }
+    };
+}
+
+/// Faster syntax for calling `verify_*` functions.
+#[macro_export]
+macro_rules! verify {
+    ( $verif: ident . $fun: ident, $( $arg: expr ),+ ) => ({
+        let mut errors = $crate::verifier::VerifierErrors::default();
+        let _ = $verif.$fun( $( $arg ),+, &mut errors);
+
+        if errors.0.is_empty() {
+            Ok(())
+        } else {
+            Err(errors)
+        }
+    });
```

Comment:
Could this be simplified to something like
```rust
$verif.$fun( $( $arg ),+, &mut errors).map_err(|_| errors)
```
?

Reply:
Indeed, I'm still paranoid about using `.map_err` in macros when raw syntax could be used, but I suppose the compiler will inline it correctly.

Reply:
Nevermind, I actually want to return an error if an error was reported (`$verif.$fun` might return `Ok` and report errors anyway).